### PR TITLE
Merge ProxyConfig Root Certs with the CSR Root Cert 

### PR DESCRIPTION
**Pull request by shankgan**
Istio-Agent side changes to merge the ProxyConfig Root Certs with the CSR Root Cert
Builds on https://github.com/istio/istio/pull/30380
Refer to [RFC](https://docs.google.com/document/d/1EFMQZNlyImLP0k0HIyoBINntchqeUq7QQgc9O_jDwZ8/edit#heading=h.io2z16q0yjxj)
Broken down into multiple commits - 
https://github.com/istio/istio/pull/30871
https://github.com/istio/istio/pull/30862
https://github.com/istio/istio/pull/31009

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: security/pkg/nodeagent/cache/secretcache.go
Comment (myidpt) [inline security/pkg/nodeagent/cache/secretcache.go:273]: I don't think we should remove this. Instead, we should in addition make sure the root of trust we construct is always based on two sources.

Before we send the root SDS response, I think we need to make sure 1) the root from the CSR response is received 2) the root list from the XDS sidechannel is received (may be empty). Otherwise, the response will be partial and incorrect.

Comment (shankgan) [reply security/pkg/nodeagent/cache/secretcache.go:273]: The code above handles what happens if the workload certificate response is present in the cache (both are merged). The code below handles what happens if there is a cache miss - we trigger and wait for the CSR response and merge with the proxyConfig certs. 
Most importantly, when the ProxyConfigCerts are updated for any reason, we trigger a callback that forces  a ROOTCA request from envoy

File: pilot/pkg/bootstrap/trustbundle.go
Comment (howardjohn) [inline pilot/pkg/bootstrap/trustbundle.go:27]: isn't this already defined in mesh config? do we need to create our own enum?

File: pilot/pkg/bootstrap/trustbundle.go
Comment (howardjohn) [inline pilot/pkg/bootstrap/trustbundle.go:42]: I am not sure why we need this (or all the associate go routines, periodic waiters, etc). is it possible to just call some `Update(TrustAnchorUpdate)` function on TrustBundle whenever the trust bundle changes?

File: pilot/pkg/bootstrap/server.go
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:1228]: 2 comments here:
* 1hr seems like a *really* long time to have to wait for configuration to update
* I am not sure we need to poll here. Can we not have the code that is handling the updates just call some callback or similar?

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pilot/pkg/bootstrap/trustbundle.go
Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go:27]: We don't have an explicit enum in the meshConfig. Also, in addition to the certs and spiffe endpoints configured in MeshConfig, we would also also need to incorporate Citadel root-certs as a source. Better to have our own enum

File: pilot/pkg/bootstrap/server.go
Comment (shankgan) [reply pilot/pkg/bootstrap/server.go:1228]: Makes sense. Given that the CA would immediately start signing certs with the new CA, the update to the TrustBundle has to happen as soon as root-certificate is rotated. In fact, the CA needs to receive an ack from all proxies before it can start using the new root of trust to sign certificates. Will add this to some callback.

File: pilot/pkg/bootstrap/trustbundle.go
Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go:42]: There are multiple go-routines that can change the TrustBundle -> the ones that (eventually) monitor Spiffe endpoints, the callback from MeshConfig Updates, the Callback from CA updates etectera. All of these eventually modify what is essentially a string of certs that need to get pushed to the proxy. Given that the trustbundle is the shared object, we want to prevent these callbacks (some of which may be very time sensitive) to wait on a write lock. I presumed a good strategy would be to serialize any writes through a single goroutine -> This is why all modifiers enqueue their update request and the processUpdates function serializes the updates.

Comment (howardjohn) [reply pilot/pkg/bootstrap/trustbundle.go:42]: it seems simpler to just have the (theoretical) `Update` function add locks?
like

```go
func (tb *TrustBundle) Update() {
tb.mutex.Lock()
defer tb.mutex.Unlock()
...update sourceConfig/mergedCerts...
}
```

Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go:42]: It would be simpler, but then the goroutine that calls the Update function *could* get locked. For instance, when MeshConfig changes, we add our callback as one of the MeshHandlers. There are other Meshhandlers (a queue of function callbacks), all of which are called within the context of the same goroutine. 
```

	for _, h := range handlers {
		h()
	}
```
Locking here would stall all the other callbacks that need to be invoked. In the current design, the goroutine that processes updates has the explicit job of just updating the trustbundle. Stalling it is acceptable

Comment (howardjohn) [reply pilot/pkg/bootstrap/trustbundle.go:42]: The current design *also* stalls. A write on an unbounded channel is blocking. Also, the time `Update` takes is either O(microsecond) or something is very wrong. Locks are used all of the place, I think we only need to be concerned about this issue if the work being done is something that may not terminate in a reasonable time.

Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go:42]: Edit: Sorry, there was code I had not pushed that increased the length of the update channel so that writers writing into the channel would not block.

Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go:42]: While the Update currently may not take much time, if we increase the size of the callback (to do verification and other activities) it could affect the time taken in the callback - definitely not longer than microseconds, but still. Its worth doing all this within a seperate goroutine IMO. Is the added complexity not worth it in your opinion?

Comment (howardjohn) [reply pilot/pkg/bootstrap/trustbundle.go:42]: It is not worth it at all in my opinion. A callback pattern is very standard both in Istio and other code base (javascript :slightly_smiling_face: ). We use it in far more critical code paths (and slower) and have never had issues with this.

This code will be invoked probably 1x at startup and never again, we don't need to prematurely optimize this

Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go:42]: Callbacks are fairly standard. However, I have spent a lot of time debugging issues in the past where a timer callback took too much time and delayed all the other callbacks in the same bucket. The ability to offload the time-sensitive part of the callback to a goroutine is a key advantage of golang. Perhaps there is some hidden cost to using the goroutine here that I am not aware of? Either way, will remove it for now.

File: pilot/pkg/bootstrap/trustbundle.go
Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go]: done

File: pilot/pkg/trustbundle/trustbundle.go
Comment (myidpt) [inline pilot/pkg/trustbundle/trustbundle.go:151]: Can we copy the value not the reference?

File: pilot/pkg/trustbundle/trustbundle.go
Comment (myidpt) [inline pilot/pkg/trustbundle/trustbundle.go:163]: Why do we need to enqueue here? A simple blocking on the lock should be OK.

File: security/pkg/pki/ca/ca.go
Comment (myidpt) [inline security/pkg/pki/ca/ca.go:410]: Can we revert the changes on this and selfsignedcarootcertrotator.go?
I don't think we need to dig so much. Consuming Citadel root via the root config is only during the migration, and a rotation normally happens at most once per year.

File: pilot/pkg/bootstrap/trustbundle.go
Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go]: Adding this in a later PR

File: pilot/pkg/bootstrap/trustbundle.go
Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go]: Will reword

File: pilot/pkg/bootstrap/trustbundle.go
Comment (myidpt) [reply pilot/pkg/bootstrap/trustbundle.go]: If the callback is only on the root rotation work, can we just remove that callback because it's a corner case?
If the merge function is the contention part, I don't think it's going to be costly :)
Shankar I understand you want to write high-quality code but the queuing thing is harder to maintain than a simple lock.

Comment (shankgan) [reply pilot/pkg/bootstrap/trustbundle.go]: Sure. Removed update goroutine and expanding scope of locks

**Comment by apolovov** [comment]
Hi, guys!
I've just tested this branch trying to create a multi-mesh federation.
### Setup:
* Two clusters with different Root CAs (their clusterDomain and trustDomain):
  * `a.local`
  * `b.local`
* In both clusters option `caCertificates` configured to remote Root CA:
  * `a.local`: 
```
data:
  mesh: |-
    trustDomain: a.local
    caCertificates:
    - pem: "-----BEGIN CERTIFICATE-----\nRoot_CA_of_b.local\n-----END CERTIFICATE-----\n"
```
  * `b.local`:
```
data:
  mesh: |-
    trustDomain: b.local
    caCertificates:
    - pem: "-----BEGIN CERTIFICATE-----\nRoot_CA_of_a.local\n-----END CERTIFICATE-----\n"
```
* in both clusters configured `PILOT_SKIP_VALIDATE_TRUST_DOMAIN`.
* in both clusters configured STRICT mTLS by PeerAuthentication and DestinationRule.
* in both clusters installed applications:
  * `a.local`: `sleep` and `httpbin` in ns `foo`.
  * `b.local`: `sleep` and `httpbin` in ns `bar`.
* in `a.local` there is `ServiceEntry` which points service `httpbin.bar.svc.b.local` to remote ingress gateway.

### Problems
* the ROOTCA secret in sidecar-proxy is being rendered wrong if I don't put `\n` in the end of my pem:
```
-----BEGIN CERTIFICATE-----
Root_CA_of_b.local
-----END CERTIFICATE----------BEGIN CERTIFICATE-----
Root_CA_of_a.local
-----END CERTIFICATE-----
``` 
* the cluster `b.local` from cluster `a.local` becomes reachable 🎉 : 
```
/ # curl httpbin.bar.svc.cluster.local:8000 -I
HTTP/1.1 200 OK
server: envoy
date: Sun, 14 Feb 2021 21:09:17 GMT
...
```
* but local services become unreachable:
```
/ # curl httpbin:8000; echo
upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: TLS error: 67108971:RSA routines:OPENSSL_internal:BLOCK_TYPE_IS_NOT_01 67109000:RSA routines:OPENSSL_internal:PADDING_CHECK_FAILED 184549382:X.509 certificate routines:OPENSSL_internal:public key routines 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
```

**Comment by apolovov** [comment]
Hi again!
I see a problem.
The envoy isn't supporting multiple certificates in UpstreamTLSContext.
https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/ssl#certificate-selection
Only DownstreamTLSContext supports.

**Comment by apolovov** [comment]
> Hi again!
> I see a problem.
> The envoy isn't supporting multiple certificates in UpstreamTLSContext.
> https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/ssl#certificate-selection
> Only DownstreamTLSContext supports.

It is a limitation of the certificate presented rather than the trust anchors themselves. This opinion of @shankgan inspired me to investigate deeper and I figured out that my problems are in badly signed Root CAs. They had the same Subject: field...
Thank you!

P.S. I can still see the `-----END CERTIFICATE----------BEGIN CERTIFICATE-----` issue.

**Comment by shankgan** [comment]
All relevant fixes associated with this pull request have been separately merged

**Pull request closed without merging**
